RHIDP-13141: Custom review pages in Orchestrator (#2136)
redhat-developer/red-hat-developers-documentation-rhdh: modules/extend_orchestrator-in-rhdh/proc-build-custom-review-pages-for-workflows.adoc

:_mod-docs-content-type: PROCEDURE

[id="build-custom-review-pages-for-workflows_{context}"]
= Build custom review pages for workflows

[role="_abstract"]

> To build a custom review page that displays workflow data in your preferred layout, adds validation messages, or integrates with your design system, implement the `getReviewComponent()` method in your form API.

Reviewer: ```suggestion
To build a custom review page that displays workflow data in your preferred layout, shows warnings or other messaging you implement, or integrates with your design system, implement the `getReviewComponent()` method in your form API.
```

> To build a custom review page that displays workflow data in your preferred layout, shows warnings or other messaging you implement, or integrates with your design system, implement the `getReviewComponent()` method in your form API.

Reviewer 2: ```suggestion
To build a custom review page that displays workflow data in a specific layout, or integrates with a design system, you must implement the `getReviewComponent()` method in the form API.
```

.Prerequisites

* You have configured the Orchestrator plugins in your {product-short} instance.
* You have a plugin or module that implements the `OrchestratorFormApi` interface from the `orchestrator-form-api` package.
* You are familiar with React component development and TypeScript.

.Procedure

. In your plugin that implements `OrchestratorFormApi`, import the required types:
+
[source,typescript]
----
import type {
  OrchestratorFormApi,
  ReviewComponentProps,
} from '@red-hat-developer-hub/backstage-plugin-orchestrator-form-api';
----

. Import the helper utilities from the `orchestrator-form-react` package:
+
[source,typescript]
----
import {
  generateReviewTableData,
  schemaHasUiHiddenFields,
  ReviewHiddenParametersAlert,
  NestedReviewTable,
} from '@red-hat-developer-hub/backstage-plugin-orchestrator-form-react';
----
+

> These utilities ensure your custom review page handles hidden fields, password masking, and nested data structures correctly.

Reviewer 2: ```suggestion
These utilities manage handles hidden fields, password masking, and nested data structures in your custom review page.
```

        />
      )}

      <NestedReviewTable data={reviewData} />

      <Box sx={{ mt: 2, display: 'flex', gap: 1 }}>
        <Button onClick={handleBack} disabled={busy}>
          Back
        </Button>
        <Button
          variant="contained"
          onClick={handleExecute}
          disabled={busy}
        >
          Execute Workflow
        </Button>
      </Box>
    </Box>
  );
};
----

. Add the `getReviewComponent()` method to your `OrchestratorFormApi` implementation:
+
[source,typescript]
----
export class MyFormApi implements OrchestratorFormApi {
  getReviewComponent() {
    return CustomReviewPage;
  }

  // ... other OrchestratorFormApi methods
}
----

. Register your custom form API with the Orchestrator plugin according to your plugin's extension mechanism.

.Verification

. Navigate to the Orchestrator plugin in the {product-short} web interface.
. Select a workflow and fill the workflow form.
. Proceed to the review step.
. Verify that your custom review page component displays with your custom layout and styling.
. Click *Back*.
. Verify that the workflow form is displayed.
. Click *Execute Workflow*.

> . Verify that the workflow runs successfully.

Reviewer 2: ```suggestion
. Open the Orchestrator plugin in the {product-short} web interface.
. Select a workflow and complete the workflow form.
. Proceed to the review step.
. Confirm that your custom review page displays with the correct layout and styling.
. Click *Back* and confirm that the workflow form is populated.
. Click *Execute Workflow* and verify that the workflow runs successfully.
```
